porch/pkg/git/git.go

```diff
@@ -1167,6 +1205,336 @@ func (r *gitRepository) traverseHistory(startCommit *object.Commit, cb commitCal
 	return nil
 }
 
+func (r *gitRepository) blobObject(h plumbing.Hash) (*object.Blob, error) {
+	return r.repo.BlobObject(h)
+}
+
+// StoreBlob is a helper method to write a blob to the git store.
+func (r *gitRepository) storeBlob(value string) (plumbing.Hash, error) {
+	data := []byte(value)
+	eo := r.repo.Storer.NewEncodedObject()
+	eo.SetType(plumbing.BlobObject)
+	eo.SetSize(int64(len(data)))
+
+	w, err := eo.Writer()
+	if err != nil {
+		return plumbing.Hash{}, err
+	}
+
+	if _, err := w.Write(data); err != nil {
+		w.Close()
+		return plumbing.Hash{}, err
+	}
+
+	if err := w.Close(); err != nil {
+		return plumbing.Hash{}, err
+	}
+
+	return r.repo.Storer.SetEncodedObject(eo)
+}
+
+func (r *gitRepository) getTree(h plumbing.Hash) (*object.Tree, error) {
+	return object.GetTree(r.repo.Storer, h)
+}
+
+func (r *gitRepository) storeTree(tree *object.Tree) (plumbing.Hash, error) {
+	eo := r.repo.Storer.NewEncodedObject()
+	if err := tree.Encode(eo); err != nil {
+		return plumbing.Hash{}, err
+	}
+
+	treeHash, err := r.repo.Storer.SetEncodedObject(eo)
+	if err != nil {
+		return plumbing.Hash{}, err
+	}
+	return treeHash, nil
+}
+
+func (r *gitRepository) UpdateDraftResources(ctx context.Context, draft *gitPackageDraft, new *v1alpha1.PackageRevisionResources, change *v1alpha1.Task) error {
```

Comment:
I assume you moved this here because we need to do multiple operations ideally under the same mutex.  I believe we probably should be doing optimistic concurrency with a compare and swap against a tag, the problem is that OCI doesn't support that.  I think this is fine, but we probably do want to split out a "pure git" layer as you suggested.

Reply:
I moved this functionality to the `gitRepository` so that the `commitHelper` is only used from within `gitRepository` functions. That means we don't need to add synchronization to the `commitHelper` functions, as we know it will only be called when holding the mutex. So it is not about doing multiple git operations under the same mutex, but just make sure that this doesn't happen concurrently with other operations on the same go-git repository.